File: arrays/next-permutation.cpp
vector<int> Solution::nextPermutation(vector<int> &A) {
    int n = A.size();
    for(int i=n-1; i>0; i--){
        if(A[i-1] < A[i]){
            reverse(A.begin()+i, A.end());
            
            int j=i;
            while(A[i-1] > A[j]) j++;
            swap(A[i-1], A[j]);
            
            return A;
        }
    }
    reverse(A.begin(), A.end());
    return A;
}

File: arrays/next-permutation.py
class Solution:
    def nextPermutation(self, A):
        i = len(A) - 1
        while i >= 0 and A[i-1] > A[i]: i -= 1

        j = len(A) - 1
        while j >= i and A[i-1] > A[j]: j -= 1
        A[i-1], A[j] = A[j], A[i-1]

        A[i:] = sorted(A[i:])
        return A

